### ISSUE TITLE
Possible evaluation leakage

### ISSUE THREAD

**Issue by chakasband** [GSanShui/XPert#2]
Hi, thanks for open-sourcing XPert.

I noticed a possible evaluation leakage issue in the training pipeline for the cold-drug/cold-cell splits. In utils.py, when a split has no valid set, the code sets val_data = test_data. Then in train_xpert.py, early stopping is performed every epoch using val_loss4, where loss4 is pcc_loss_sum(deg_output, trt_raw_data - ctl_raw_data), i.e. a PCC-based loss on the predicted differential expression (DEG / treatment-control difference).

At evaluation time, the reported Pearson_deg is also computed on the same differential expression target (y_deg = y - ctl, f_deg = f - ctl). So for these splits, it seems that:

validation == test
**early stopping is done on the test fold**
and the stopping criterion is directly tied to the final Pearson_deg metric 
Could you clarify whether this was the intended protocol used for the paper results, or whether there should have been a separate validation split inside the training fold?

**Comment by GSanShui** [comment]
Hi, thank you for pointing this out. You are right. But this was an oversight during the code simplification for open-sourcing.

To clarify, the results reported in our paper are strictly isolated and free from this leakage. Our actual pipeline was as follows:

- **Hyperparameter Tuning**: We used a **3:1:1 train/val/test split** to optimize hyperparameters based on the validation set.
- **5-Fold Cross-Validation**: With fixed parameters, we performed a standard 5-fold CV. For each fold, instead of using the test set, **we selected 10% of the training set based on the split attributes** to serve as an internal validation set for early stopping. The test set remained strictly isolated.

We apologize for the confusion, but please be assured the paper's reported results are unaffected. 

We are currently updating and testing the code to reflect this exact pipeline and will push the fix soon. 

In the meantime, please refer to the code snippet in the `utils.py` below for the fix:

```
from sklearn.model_selection import train_test_split

# ......

tr_data = data[data.obs[nfold] == 'train']
val_data = data[data.obs[nfold] == 'valid']
test_data = data[data.obs[nfold] == 'test']

# for five-fold cross-validation
if val_data.n_obs == 0:
    if 'cold_drug' in nfold:
        unique_attrs = tr_data.obs['pert_id'].unique()
        _, val_attrs = train_test_split(unique_attrs, test_size=0.1, random_state=42)
        is_val = tr_data.obs['pert_id'].isin(val_attrs)
        
    elif 'cold_cell' in nfold:
        # Cold-cell
        unique_attrs = tr_data.obs['cell_iname'].unique()
        _, val_attrs = train_test_split(unique_attrs, test_size=0.1, random_state=42)
        is_val = tr_data.obs['cell_iname'].isin(val_attrs)
        
    else:
        indices = np.arange(tr_data.n_obs)
        _, val_indices = train_test_split(indices, test_size=0.1, random_state=42)
        is_val = np.zeros(tr_data.n_obs, dtype=bool)
        is_val[val_indices] = True

    val_data = tr_data[is_val].copy()
    tr_data = tr_data[~is_val].copy()
    # =======================================================================

logger.info(f'train data:{tr_data.n_obs}')  
logger.info(f'Valid data:{val_data.n_obs}')
logger.info(f'Test data:{test_data.n_obs}')
```

Thanks again for your valuable feedback!

**Comment by GSanShui** [comment]
Hi,

Thank you for reaching out and for your detailed feedback on XPert!

We received your email regarding the potential evaluation leakage. Since
you also raised this question on our GitHub repository, we have posted a
detailed explanation of our actual evaluation methodology, along with the
code snippet for the fix, directly in the GitHub thread.

You can check out our complete response here:
https://github.com/GSanShui/XPert/issues/2

Please feel free to continue the conversation on GitHub if you have any
follow-up questions. We really appreciate your valuable input!

Best regards,

Yue Guo

chakasband ***@***.***> 于2026年3月9日周一 00:11写道：

**Comment by chakasband** [comment]
Great, thank you for the quick response and for clarifying the training pipeline!!!
